File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/Query/QueryController.php
<?php

namespace App\Http\Controllers\Api\V1\Query;

use OpenApi\Attributes as OA;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\Query\Services\QueryService;
use Illuminate\Http\Request;

#[OA\Tag(name: 'Query')]
class QueryController extends ApiController
{
    public function __construct(private readonly QueryService $service) {}

    public function index(Request $request)
    {
        return $this->success($this->service->listProducts((int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function show(int $productId)
    {
        return $this->success($this->service->showProduct($productId));
    }

    public function sync(Request $request, int $productId)
    {
        return $this->success($this->service->sync($request->user(), $productId));
    }

    public function rebuild(Request $request)
    {
        return $this->success($this->service->rebuild($request->user()));
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/query/QueryController.java
package com.pbshop.java.spring.maven.jpa.postgresql.query;

import java.util.Map;

import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.api.ApiResponse;

@RestController
@RequestMapping("${pbshop.api.base-path:/api/v1}")
public class QueryController {

    private final QueryService queryService;

    public QueryController(QueryService queryService) {
        this.queryService = queryService;
    }

    @GetMapping("/query/products")
    public ApiResponse<Map<String, Object>> listProducts(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit
    ) {
        return ApiResponse.success(queryService.listProducts(page, limit));
    }

    @GetMapping("/query/products/{productId}")
    public ApiResponse<Map<String, Object>> showProduct(@PathVariable Long productId) {
        return ApiResponse.success(queryService.showProduct(productId));
    }

    @PostMapping("/admin/query/products/rebuild")
    public ApiResponse<Map<String, Object>> rebuild(@AuthenticationPrincipal AuthenticatedUserPrincipal principal) {
        return ApiResponse.success(queryService.rebuild(principal));
    }

    @PostMapping("/admin/query/products/{productId}/sync")
    public ApiResponse<Map<String, Object>> sync(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long productId
    ) {
        return ApiResponse.success(queryService.sync(principal, productId));
    }
}
